### PR TITLE
App Hosting emulator - package manager detection

### PR DESCRIPTION
**Pull request by mathu97**
App hosting emulator will detect the package manager and run it's dev command on `firebase emulators:start`

Supported package managers:
1. npm
2. pnpm
3. yarn

### CODE REVIEW

File: src/emulator/apphosting/serve.ts [hunk 1]
```diff
@@ -2,37 +2,55 @@
...
+  return await hostUrl;
 }
 
 /**
  * Exported for unit testing
  */
-export async function serve(options: any, port: string) {
-  // TODO: update to support other package managers and frameworks other than NextJS
-  await wrapSpawn("npm", ["run", "dev", "--", "-H", options.host, "-p", port], process.cwd());
+export function getHostUrlFromString(data: string): string | undefined {
```
Comment (taeold) [inline src/emulator/apphosting/serve.ts:51]: Can you clarify what this function is doing? It looks like it's doing few things like

1) ignoring URLs that starts with https://  (why?)
2) stripping out "/" (why?)

File: src/emulator/apphosting/serve.ts [hunk 2]
```diff
@@ -2,37 +2,55 @@
...
+   * */
+
+  const hostUrl = new Promise<string>((resolve, reject) => {
+    const serve = spawn(packageManager, ["run", "dev"], {
+      cwd: rootDir,
+    });
+
+    serve.stdout.on("data", (data: any) => {
+      process.stdout.write(data);
+      const url = getHostUrlFromString(data.toString());
```
Comment (taeold) [inline src/emulator/apphosting/serve.ts:36]: I think we can reasonably expect that the

1) dev server will run on localhost
2) dev server will respect the `PORT` environment variable

I feel like these expectation is probably as reasonable as expecting the dev server to print out the URL. wdyt?

Comment (mathu97) [reply src/emulator/apphosting/serve.ts:36]: +1 reading from STDOUT definitely does not feel right, and I think these assumptions are valid.

File: src/emulator/apphosting/serve.ts (hunk 1 above)
Comment (mathu97) [reply src/emulator/apphosting/serve.ts]: outdated - not using the approach of reading host/port from STDOUT anymore

**Review by taeold — APPROVED** [review]

**Pull request merged**